### DISCUSSION TITLE
Unable to build on Manjaro linux

### DISCUSSION THREAD

**Discussion by zubayrali** [tiddly-gittly/TidGi-Desktop#76] [Q&A]
Hello,

I cloned the repo, ran the commands as instructed but it throws an error after I run `npm run make:linux-x64` I've never worked with electron before, I couldn't find a fix after hours of googling, please help.

```javascript
An unhandled rejection has occurred inside Forge:
Error: packageJSON.main must be set to a valid entry point for your Electron app
    at _default (/home/kira/Downloads/pkgs/TiddlyGit-Desktop/node_modules/@electron-forge/core/src/api/package.ts:107:11)
    at /home/kira/Downloads/pkgs/TiddlyGit-Desktop/node_modules/@electron-forge/cli/src/electron-forge-package.ts:30:3

Electron Forge was terminated. Location:
{}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! TiddlyGit@0.4.0 package: `electron-forge package`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the TiddlyGit@0.4.0 package script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/kira/.npm/_logs/2021-07-05T14_08_30_810Z-debug.log
```

node version `v14.17.2`
npm version `7.19.1`
electron version `v13.1.5`

**Comment by linonetwo** [comment]
Hi, I'm sorry to hear that.

This is something related to electron-forge, a starter framework I use to simplify this packaging process, and it works good, I never run into this before.

You can try `npm run start` first, before you try to backage it, because package is much complex than dev for electron

**Reply by zubayrali** [reply]
Hey, thank you for the quick response

Running `npm run start` throws this error:

```javascript
An unhandled rejection has occurred inside Forge:
Error: Could not find any Electron packages in devDependencies
    at getElectronModuleName (/home/kira/Desktop/tiddlygittly/node_modules/@electron-forge/core/src/util/electron-version.ts:56:11)
    at getElectronVersion (/home/kira/Desktop/tiddlygittly/node_modules/@electron-forge/core/src/util/electron-version.ts:92:23)
    at _default (/home/kira/Desktop/tiddlygittly/node_modules/@electron-forge/core/src/api/start.ts:49:11)
    at /home/kira/Desktop/tiddlygittly/node_modules/@electron-forge/cli/src/electron-forge-start.ts:58:19

Electron Forge was terminated. Location:
{}
```

**Reply by zubayrali** [reply]
I installed `electron-forge` cli and ran `electron-forge import` and now its throwing this error

```javascript
> TiddlyGit@0.4.0 start
> electron-forge start

✔ Checking your system
✔ Locating Application
✔ Preparing native dependencies
✔ Launching Application
Error launching app
Unable to find Electron app at /home/kira/Desktop/tiddlygittly

Cannot find module '/home/kira/Desktop/tiddlygittly/.webpack/main'. Please verify that the package.json has a valid "main" entry
```

**Reply by linonetwo** [reply]
I think you can just cd into the repo folder, and run `npm i` then `npm start`, no need to "installed electron-forge cli and ran electron-forge import"

Hope you succeed this time!

**Reply by linonetwo** [reply]
I think getting `Could not find any Electron packages in devDependencies` means you forget to `npm i`?

**Reply by linonetwo** [reply]
If you followed this part, but still getting these, maybe means you failed to install electron, just delete 'electron' folder in the 'node_modules' and rerun `npm i`

![截屏2021-07-06 01 12 21](https://user-images.githubusercontent.com/3746270/124503061-55bc5a80-ddf7-11eb-8963-3e58499ab19e.png)

**Reply by linonetwo** [reply]
feel free to ask me about this, it is hard to understand a new repo.

**Reply by zubayrali** [reply]
Hey, thank you

I did run `npm i` before `npm start`, I've deleted node_modues and ran it again but it's not working, I will try to build it on another computer.

**Reply by linonetwo** [reply]
Are you in master branch? And are you running npm start after cd into the repo folder? I surely have `Electron packages in devDependencies`, but if you run `npm start` with wrong pwd, this will failed.

**Reply by linonetwo** [reply]
You need to run `npm` commands with the same pwd as `forge.config.js`

**Reply by zubayrali** [reply]
yeah, I did `cd`into `TiddlyGit-Desktop` before running the npm commands

**Reply by linonetwo** [reply]
I only have MacOS and Windows device, maybe you can try a Windows machine, to check whether this is a platform problem.

**Reply by zubayrali** [reply]
so I am trying to build it on another computer(Manjaro linux)

I change the config to:

```javascript
makers: [
    {
      name: '@electron-forge/maker-snap',
      config: {
        features: {
          audio: true,
          mpris: 'com.example.mpris',
          webgl: true
        },
        summary: 'Pretty Awesome'
      }
    }
  ],

```

I installed snapcraft binaries and ran `npm run make:linux-x64` and it threw this error:

```javascript
[masi@masi TiddlyGit-Desktop]$ npm run make:linux-x64      

> TiddlyGit@0.4.0 make:linux-x64
> cross-env NODE_ENV=production electron-forge make --platform=linux --arch=x64

✔ Checking your system
✔ Resolving Forge Config
We need to package your application before we can make it
⠋ Preparing to Package Application for arch: x64You have set packagerConfig.ignore, the Electron Forge webpack plugin normally sets this automatically.                                                                                                                 

Your packaged app may be larger than expected if you dont ignore everything other than the '.webpack' folder                        
⠇ Compiling Main Process Code(node:15598) [DEP_WEBPACK_COMPILATION_CACHE] DeprecationWarning: Compilation.cache was removed in favor of Compilation.getCache()
(Use `node --trace-deprecation ...` to show where the warning was created)
⠴ Compiling Main Process CodeCopying tiddlywiki dependency to dist
✔ Compiling Main Process Code
✔ Compiling Renderer Template
✔ Compiling Renderer Preload: main_window
✔ Preparing to Package Application for arch: x64
✔ Preparing native dependencies
✔ Packaging Application
Making for the following targets: snap
⠸ Making for target: snap - On platform: linux - For arch: x64The browser-sandbox feature will trigger a manual review in the Snap store.
⠦ Making for target: snap - On platform: linux - For arch: x64/bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
⠴ Making for target: snap - On platform: linux - For arch: x64/bin/sh: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8)
⠹ Making for target: snap - On platform: linux - For arch: x64locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_COLLATE to default locale: No such file or directory
Traceback (most recent call last):
  File "/snap/snapcraft/6596/bin/snapcraft", line 33, in <module>
    sys.exit(load_entry_point('snapcraft==4.8.3', 'console_scripts', 'snapcraft')())
  File "/snap/snapcraft/6596/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/snap/snapcraft/6596/lib/python3.6/site-packages/click/core.py", line 760, in main
    _verify_python3_env()
  File "/snap/snapcraft/6596/lib/python3.6/site-packages/click/_unicodefun.py", line 130, in _verify_python3_env
    " mitigation steps.{}".format(extra)
RuntimeError: Click will abort further execution because Python 3 was configured to use ASCII as encoding for the environment. Consult https://click.palletsprojects.com/python3/ for mitigation steps.

This system lists a couple of UTF-8 supporting locales that you can pick from. The following suitable locales were discovered: en_IN.utf8, en_US.utf8, ko_KR.utf8

Click discovered that you exported a UTF-8 locale but the locale system could not pick up from it because it does not exist. The exported locale is 'C.UTF-8' but it is not supported
✖ Making for target: snap - On platform: linux - For arch: x64

An unhandled error has occurred inside Forge:
An error occured while making for target: snap
Command failed with a non-zero return code (1):
/var/lib/snapd/snap/bin/snapcraft snap --target-arch=amd64 --output=/home/masi/Desktop/TiddlyGit-Desktop/out/make/snap/x64/tiddlygit_0.4.0_amd64.snap
Error: Command failed with a non-zero return code (1):
/var/lib/snapd/snap/bin/snapcraft snap --target-arch=amd64 --output=/home/masi/Desktop/TiddlyGit-Desktop/out/make/snap/x64/tiddlygit_0.4.0_amd64.snap
    at ChildProcess.<anonymous> (/home/masi/Desktop/TiddlyGit-Desktop/node_modules/@malept/cross-spawn-promise/src/index.ts:172:16)
    at ChildProcess.emit (events.js:375:28)
    at ChildProcess.emit (domain.js:470:12)
    at maybeClose (internal/child_process.js:1055:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)


```

**Reply by linonetwo** [reply]
Why the path is `Desktop/tiddlygittly/node_modules`? should it be like `/Users/linonetwo/Desktop/github/TiddlyGit-Desktop/node_modules` ?

**Reply by linonetwo** [reply]
There should be a package json along side node_modules 
<img width="370" alt="截屏2021-07-06 下午2 46 57" src="https://user-images.githubusercontent.com/3746270/124554908-10d10c00-de69-11eb-95f7-6c1c386cfdb2.png">

**Reply by zubayrali** [reply]
I also tried the [Flatpak maker](https://www.electronforge.io/config/makers/flatpak)
with the below config 

```javascript
{
  name: '@electron-forge/maker-flatpak',
  config: {
    options: {
      categories: ['Video'],
      mimeType: ['video/h264']
    }
  }
}
```

and it still threw this error:

```javascript
[masi@masi TiddlyGit-Desktop]$ npm run make:linux-x64                

> TiddlyGit@0.4.0 make:linux-x64
> cross-env NODE_ENV=production electron-forge make --platform=linux --arch=x64

✔ Checking your system
✔ Resolving Forge Config

An unhandled rejection has occurred inside Forge:                                                                                   
Error: Cannot make for linux and target flatpak: the maker declared that it cannot run on linux
    at _default (/home/masi/Desktop/TiddlyGit-Desktop/node_modules/@electron-forge/core/src/api/make.ts:142:13)
    at /home/masi/Desktop/TiddlyGit-Desktop/node_modules/@electron-forge/cli/src/electron-forge-make.ts:44:5

Electron Forge was terminated. Location:                                                                                            
{}

```
btw I did remember to run `npm i @electron-forge/maker-flatpak -D` before trying this.

**Reply by zubayrali** [reply]
> I only have MacOS and Windows device, maybe you can try a Windows machine, to check whether this is a platform problem.

Unfortunately I didn't try to build it on Windows, I did try the .exe which worked fine except for a few errors it gave which I couldn't understand.

Would it be possible to make a flatpak or snap? maybe an AppImage?

**Reply by zubayrali** [reply]
> Why the path is `Desktop/tiddlygittly/node_modules`? should it be like `/Users/linonetwo/Desktop/github/TiddlyGit-Desktop/node_modules` ?

I cloned the repo directly onto the Desktop directory, 
if /github is supposed to be GithubDesktop, I don't have it on my Manjaro Linux

**Reply by linonetwo** [reply]
No, I just place all cloned repo in the `github` folder, this is not mandatory.

Hope you can get it running ,and add the packgers you need. I don't know much about those linux packages.

**Reply by linonetwo** [reply]
It seems better in your another computer, you can npm start this time right?

**Reply by zubayrali** [reply]
oh I did run `npm run start` this time and it worked 

```node
cached modules 12.6 MiB [cached] 11547 modules
runtime modules 25.9 KiB 13 modules
webpack 5.38.1 compiled successfully in 11728 ms
✔ Compiling Renderer Code
No issues found.
```

although for whatever reasons no other windows seem to be working,

![image](https://user-images.githubusercontent.com/31969517/124556903-82538f00-de56-11eb-8969-13bc3bbc3366.png)

**Reply by linonetwo** [reply]
You can right click the white window and use development tool to see what happened, you can post the screenshot of errors here

**Reply by zubayrali** [reply]
Here is the console:

![image](https://user-images.githubusercontent.com/31969517/124557147-ca72b180-de56-11eb-8a34-8ea8e73ad580.png)

**Reply by linonetwo** [reply]
I update the master branch, you can try this is fixed, hopefully. I didn't encounter this in MacOS before.

**Reply by zubayrali** [reply]
Awesome! I'll try and let you know, thank you
BTW I have a lil bit of experience with web development and Reactjs, I would love to contribute, let me know if there are beginner-friendly tasks for me.

**Reply by linonetwo** [reply]
Basically, there is no much to do in the react part, most of the UI code was copy from the WebCatalog though.

But you can try https://github.com/tiddly-gittly/TiddlyGit-Desktop/projects/3#card-39985517 to make it more friendly, add "open in code editor" and "open in git GUI" button, and open vscode and github desktop by default, so debug tw state and sync state can be more easy. And this task is relatively easy.

**Reply by zubayrali** [reply]
Hey

I ran `git pull` and then did `npm i`, I still get white blank windows other than the main window

![image](https://user-images.githubusercontent.com/31969517/124588294-416b7280-de76-11eb-85ad-3c63ff67c140.png)

![image](https://user-images.githubusercontent.com/31969517/124589031-24836f00-de77-11eb-89ba-4e56f5e3bf65.png)

![image](https://user-images.githubusercontent.com/31969517/124589101-382ed580-de77-11eb-8f27-c2a5bd6f56e6.png)

**Reply by zubayrali** [reply]
> Basically, there is no much to do in the react part, most of the UI code was copy from the WebCatalog though.
> 
> But you can try https://github.com/tiddly-gittly/TiddlyGit-Desktop/projects/3#card-39985517 to make it more friendly, add "open in code editor" and "open in git GUI" button, and open vscode and github desktop by default, so debug tw state and sync state can be more easy. And this task is relatively easy.

Will definitely look into it, thank you
I will work on it as soon as I get it working on my main machine :)

**Reply by zubayrali** [reply]
Would you like to see the output from the webpack running on localhost:9000 ?

**Reply by linonetwo** [reply]
This is quite similar to https://github.com/tiddly-gittly/TiddlyGit-Desktop/issues/56  but I can't reproduce on my side...

**Reply by linonetwo** [reply]
In the code, there definitely have FolderIcon defined. My npm start won't give me this,  can you click on this link and see what happened in that line?

![截屏2021-07-06 下午7 50 27](https://user-images.githubusercontent.com/3746270/124595294-6b7f5d80-de93-11eb-8afb-b44dce6d5cc4.png)

**Reply by zubayrali** [reply]
hey sorry i got busy with my day job, I'll get back to this once I'm done with work.

thank you

**Reply by zubayrali** [reply]
clicked on NewWikiForm.tsx 

it gave this `Uncaught ReferenceError: FolderIcon is not defined` error

![image](https://user-images.githubusercontent.com/31969517/124721287-9bc40c00-df26-11eb-979d-d285ce93ab42.png)

**Reply by zubayrali** [reply]
btw the Notifications window is working

**Reply by linonetwo** [reply]
I think this is caused by the webpack, it didn't handle the import of that icon properly. I can't reproduce on my side...

**Reply by zubayrali** [reply]
Okay, I'll try to figure it out.

**Reply by linonetwo** [reply]
Try comment out these lines https://github.com/tiddly-gittly/TiddlyGit-Desktop/blob/6791b66d71bc01b58e7319584161a24cfc82e65c/webpack.rules.js#L48-L67

**Reply by linonetwo** [reply]
I think my config of ts import tree shaking is somehow wrong. So just disable them may fix this.

**Reply by zubayrali** [reply]
Yay! it worked

**Reply by linonetwo** [reply]
I don't know if TiddlyGit works in your machine, I don't have a Linux machine to test. But AFAIK it still have some bug in Windows...People can't create and open a new workspace.

**Reply by zubayrali** [reply]
Yes, i am on the Add workspace window and these tabs don't seem to be working.

**Reply by zubayrali** [reply]
![image](https://user-images.githubusercontent.com/31969517/124723956-29a0f680-df29-11eb-8c59-dabb42c8c408.png)

**Reply by zubayrali** [reply]
> I don't know if TiddlyGit works in your machine, I don't have a Linux machine to test. But AFAIK it still have some bug in Windows...People can't create and open a new workspace.

you could test on a virtual machine 😅

**Reply by linonetwo** [reply]
oh, probably you need to comment out all these tsImportPluginFactory, they seem to only work properly in MacOS, which shouldn't be, these are JavaScript...

**Reply by zubayrali** [reply]
Hey sorry for abandoning this, I got busy with some work-related stuff, I'll try debugging this in my free time, thank you

**Reply by linonetwo** [reply]
Ok, I've update many things, so maybe bug is already gone. Just work on your job first, its totally fine.
